### PR TITLE
Feature/article tag rule

### PR DESCRIPTION
**Pull request by todd-dembrey**
@tm-kn 

Adds the ability to filter by tags on articles

### CODE REVIEW

**Comment by tm-kn** [comment]
I think we might need to put validation errors as `ugetext`.

File: molo/surveys/adapters.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            date_to = datetime.datetime.max
+
+        tag_visits = self.request.session.setdefault(
+            'tag_count',
+            defaultdict(dict),
+        )
+
+        visits = tag_visits.get(str(tag.id), dict())
+        valid_visits = [visit for visit in visits.values()
+                        if date_from <= parse_datetime(visit) <= date_to]
```
Comment (tm-kn) [inline molo/surveys/adapters.py]: Are you sure there's no possibility of `parse_datetime` returning `None` and that `visit` will always be a valid string? We don't want it crashing on users.

File: molo/surveys/rules.py [hunk 2]
```diff
@@ -231,3 +235,92 @@ def test_user(self, request):
...
+
+class ArticleTagRule(AbstractBaseRule):
+    EQUALS = 'eq'
+    GREATER_THAN = 'gt'
+    LESS_THAN = 'lt'
+
+    OPERATORS = {
+        EQUALS: lambda a, b: a == b,
+        GREATER_THAN: lambda a, b: a > b,
+        LESS_THAN: lambda a, b: a < b,
```
Comment (tm-kn) [inline molo/surveys/rules.py]: 👍 like how readable the lambdas are.

File: molo/surveys/adapters.py (hunk 1 above)
Comment (tm-kn) [inline molo/surveys/adapters.py]: I presume `date_from` and `date_to` will have some timezone attached to them if inputted in the admin site. And the current `visit` will always be in UTC and comparison should go smoothly between different timezones? You might want to test that in the unit tests if you do not already do so.

File: molo/surveys/adapters.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class SurveysSegmentsAdapter(SessionSegmentsAdapter):
+    def add_page_visit(self, page):
+        super(SurveysSegmentsAdapter, self).add_page_visit(page)
+        tag_visits = self.request.session.setdefault(
+            'tag_count',
+            defaultdict(dict),
+        )
+        if isinstance(page, ArticlePage):
```
Comment (tm-kn) [inline molo/surveys/adapters.py]: I am not sure whether `page` will be `specific`? you might want to test it.

Comment (todd-dembrey) [reply molo/surveys/adapters.py]: Probably safer, good idea.

File: molo/surveys/adapters.py (hunk 1 above)
Comment (todd-dembrey) [reply molo/surveys/adapters.py]: `date_from` and `date_to` are not timezone aware. The visits are currently logged based on server time so adding time zones into the admin would add confusion. However this is not ideal as we would want to know the users timezone. Perhaps this should be noted somewhere.

Comment (tomdyson) [reply molo/surveys/adapters.py]: > Perhaps this should be noted somewhere

Please do note this in the developer docs.

**Pull request merged**